core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -979,6 +1004,21 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         }
     }
 
+    private void verifyPrimary() {
+        if (shardRouting.primary() == false) {
+            // must use exception that is not ignore be replication logic. See TransportActions.isShardNotAvailableException
```

Comment:
// must use exception that is not **ignored by** replication logic. (also 2 more occurrences of this in IndexShard)